### ISSUE TITLE
Bug: Attribute restSearch with excluding decayed attributes is abnormally slow

### ISSUE THREAD

**Issue by Tahantos** [MISP/MISP#8660]
### Expected behavior

Using restSearch to export attributes while excluding decayed attributes should not take multiple times longer than exporting all attributes.

### Actual behavior

When using restSearch with the following options:

`
{
    "returnFormat": "text",
    "page": "1",
    "limit": "250000",
    "type": "url"
}
`

The request duration is: 63692.06 ms

After adding the option "excludeDecayed": 1, the request duration jumps to: 612608.56 ms

This happens with other attribute types as well.

I have tested this out in 2 separate but similar environments.

### Steps to reproduce

1. Have any kind of decaying model enabled for any attribute you want to test with
2. Do a restSearch with any means possible with the excludeDecayed disabled
3. Do another restSearch with the excludeDecayed enabled
4. Compare results

### Version

2.4.163

### Operating System

Debian

### Operating System version

11

### PHP version

7.4

### Browser

Chrome

### Browser version

106.0.5249.103

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by advbarbieux** [comment]
Facing the same issue in a project involving MISP at my company.

Further investigation on our end has led us to the following conclusions.

### 1. Specifying any parameter requiring decay score computation causes this performance issue.
This includes the following parameters:
- `includeDecayScore` (both when set to `true` and when set to `false`);
- `excludeDecayed`.

On a local instance of MISP (based on the [Docker image](https://github.com/MISP/misp-docker)) running on my machine with attributes fed from the _CIRCL OSINT Feed_ default feed (277,934 attributes), the `POST /attributes/restSearch` API yielded the following results upon requesting the first 10,000 attributes (`{"page": 1, "limit": 10000}`):

**Test 1: no additional parameters**
Attributes are retrieved in **~0.27 secs**.

**Test 2: specifying `"includeDecayScore": false`**
Attributes are retrieved in **~3.12 secs**.
However, according to [the docs](https://www.misp-project.org/openapi/#tag/Attributes/operation/restSearchAttributes), `false` is the default value so test 2 should be equivalent to test 1, yet we're experiencing a ~×11 slowdown here.

**Test 3: specifying `"includeDecayScore": true`**
Attributes are retrieved in **~3.12 secs** as well.
It appears this triggers the same behavior as test 2.

**Test 4: specifying `"excludeDecayed": false`**
Attributes are retrieved in **~3.12 secs** as well.
Same behavior as test 2 and 3, most likely, which is unexpected because `false` is the default value for the `excludeDecayed` parameter (test 4 should then be equivalent to test 1 and 2).

**Test 5: specifying `"excludeDecayed": true`**
Attributes are retrieved in **~3.12 secs** as well.
Same behavior as test 2, 3 and 4, apparently.

### 2. This issue could be solved by performing bulk DB queries when requesting decay-related info about attributes.
Looking into the source code, we figured out the following:
- Attributes returned by the API are fetched by the [`Attribute::fetchAttributes`](https://github.com/MISP/MISP/blob/2.4/app/Model/Attribute.php#L1697) function;
- At [`Attribute.php:1889`](https://github.com/MISP/MISP/blob/2.4/app/Model/Attribute.php#L1889), a `do`/`while` loop iterates over all fetched attributes;
- In this loop, at [line 1953](https://github.com/MISP/MISP/blob/2.4/app/Model/Attribute.php#L1953) a function named [`DecayingModel::attachScoresToAttribute`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModel.php#L609) is called;
- At [`DecayingModel.php:625`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModel.php#L625), this function in turn calls a function named [`DecayingModel::getScore`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModel.php#L687);
- At [`DecayingModel.php:703`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModel.php#L703), this function in turn calls a function named [`DecayingModelBase::computeCurrentScore`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModelsFormulas/Base.php#L129);
- At [`DecayingModelsFormulas/Base.php:136`](https://github.com/MISP/MISP/blob/2.4/app/Model/DecayingModelsFormulas/Base.php#L136), a function named [`Sighting::listSightings`](https://github.com/MISP/MISP/blob/2.4/app/Model/Sighting.php#L921) is called;
- The [`Sighting::listSightings`](https://github.com/MISP/MISP/blob/2.4/app/Model/Sighting.php#L921) function performs the following actions:
  - Retrieve the attribute again at [`Sighting.php:928`](https://github.com/MISP/MISP/blob/2.4/app/Model/Sighting.php#L928), which can lead to a new DB query;
  - Retrieve the attribute's sightings at [`Sighting.php:957`](https://github.com/MISP/MISP/blob/2.4/app/Model/Sighting.php#L957) through a new DB query.

**TL;DR the code as-is performs 2 DB queries for each attribute previously retrieved. So making an API call to retrieve 10,000 attributes will result in ~20,000 DB queries.**

This causes massive pressure on the DB backend, which is most likely the bottleneck causing this issue.

This issue could therefore be solved by performing bulk DB queries (i.e. a single query phrased as "_find sightings linked to this list of attribute IDs_") instead of performing a DB query targeting a single attribute ("_find sightings linked to this specific attribute ID_") for each attribute.

**Comment by adulau** [comment]
@mokaddem the test from @advbarbieux sounds quite interesting.

**Comment by mokaddem** [comment]
Hello @advbarbieux and thanks for your tests.
I wasn't able to reproduce the result you have on **Test 2**. For me, it gives the same result as **Test 1**.
However your analysis helped me fixed some major performance issues.
I've pushed ec769c3f2714f9d90e3a72d095515d8eb0e813b7 and c797865c7c9f69ce4f0160492660eb2ac030918d on `develop` which patch the following:
- Stopped collecting the same decaying models from the DB over and over by using a cache
- Only fetch the last sighting (instead of all of them).

With these in place, I managed to improve the request duration by a factor of 5; which is now much faster than requesting all sightings with the `includeSightings` options. This factor can be even higher if you have fine-tuned your DB.

Let us know if you stop other ways to improve the performance of the application!
